pkg/distributor/distributor.go

```diff
@@ -1028,6 +1037,61 @@ func (d *Distributor) prePushForwardingMiddleware(next push.Func) push.Func {
 	}
 }
 
+// prePushEphemeralMiddleware is used as push.Func middleware in front of push method.
+// If marking series as ephemeral is enabled, this middleware uses the ephemeral series
+// provider to determine whether a time series should be marked as ephemeral.
+func (d *Distributor) prePushEphemeralMiddleware(next push.Func) push.Func {
+	if !d.cfg.EphemeralMetricsEnabled {
+		return next
+	}
+
+	return func(ctx context.Context, pushReq *push.Request) (*mimirpb.WriteResponse, error) {
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		req, err := pushReq.WriteRequest()
+		if err != nil {
+			return nil, err
+		}
+
+		// No ephemeral series from ruler.
+		if req.Source == mimirpb.RULE {
+			return next(ctx, pushReq)
+		}
+
+		req.EphemeralTimeseries = nil
+
+		ephemeralChecker := d.EphemeralCheckerByUser.EphemeralChecker(userID)
+		if ephemeralChecker != nil {
+			req.EphemeralTimeseries = mimirpb.PreallocTimeseriesSliceFromPool()
+
+			var deleteTs []int
+			for ix := 0; ix < len(req.Timeseries); ix++ {
+				ts := req.Timeseries[ix]
+
+				if !ephemeralChecker.IsEphemeral(ts.Labels) {
```

Comment:
If we pass `req.Source` to `IsEphemeral`, then we can have checker decide whether series originated from ruler can be marked as ephemeral or not. That's probably better than having `if req.Source == mimirpb.RULE` check above. (Can be in separate PR)

Reply:
Alternatively, we can pass the source to `EphemeralChecker` call. That would allow us to have a two separate sets of configs, but if someone wants the same config for API and Ruler, then they would need to duplicate them.

Reply:
Interesting idea, I like it. So then the list of label matchers would have to be keyed by the store, kind of like:

```
overrides:
  anonymous:
    ephemeral_series_matchers:
      api:
        - __name__="metric_from_api"
      ruler:
        - __name__="metric_from_ruler"
```

Reply:
I implemented this in b8e8c76ed0c5ae04350f35b97f1b935a7d711616